src/Tasks/Microsoft.NET.Build.Tasks/ReferenceInfo.cs

```diff
@@ -54,17 +54,33 @@ public static IEnumerable<ReferenceInfo> CreateReferenceInfos(IEnumerable<ITaskI
         public static IEnumerable<ReferenceInfo> CreateDirectReferenceInfos(
             IEnumerable<ITaskItem> referencePaths,
             IEnumerable<ITaskItem> referenceSatellitePaths,
-            bool projectContextHasProjectReferences,
+            LockFileLookup lockFileLookup,
             Func<ITaskItem, bool> isRuntimeAssembly)
         {
-
-            bool filterOutProjectReferenceIfInProjectContextAlready(ITaskItem referencePath)
+            bool lockFileContainsProject(ITaskItem referencePath)
             {
-                return (projectContextHasProjectReferences ? !IsProjectReference(referencePath) : true);
+                if (lockFileLookup == null)
+                {
+                    return false;
+                }
+
+                if (!IsProjectReference(referencePath))
+                {
+                    return false;
+                }
+
+                string outputName = Path.GetFileName(referencePath.ItemSpec);
+                string projectName = Path.GetFileNameWithoutExtension(outputName);
+                if (string.IsNullOrEmpty(projectName))
+                {
+                    return true;
+                }
+
+                return lockFileLookup.GetProject(projectName) != null;
```

Comment:
I don't know what will happen - it will probably fail to find it.  That was something I was wondering about yesterday.  Do you know a reliable way to get the project name for a given "ITaskItem" that represents a project reference?

Reply:
I went through a period of time in the PR trying to use "FusionName", but there's also [this area](https://github.com/dotnet/sdk/blob/2ea9341cf0f7c43910072002db4e218396c41f28/src/Tasks/Microsoft.NET.Build.Tasks/SingleProjectInfo.cs#L85) that tries to use MetadataKeys.MSBuildSourceProjectFile and that wasn't set in some scenarios that I saw in the various unit and functional tests.  But would that be better to use if it is present?

Reply:
I've updated this and added a test - hopefully this will cover the scenario.  :-)